tools/tpm2_ptool.py

```diff
@@ -1063,175 +1291,90 @@ def generate_options(self, group_parser):
+        pinopts.add_argument(
+            '--sopin',
+            help='The Administrator pin,'),
+        pinopts.add_argument(
+            '--userpin',
+            help='The User pin.\n'),
 
-                # Now get the secondary object from db
-                sobj = db.getsecondary(token['id'])
+    # Creates a new key
+    def new_key_create(self, args, sobjctx, sobjauth, objauth, tpm2):
+        path = args['path']
+        # create a tertiary object under the loaded secondary object
+        # Map onto the defaults in tpm2_create and store the attributes into
+        # the db.
+        alg = args['algorithm']
+        if alg.startswith('rsa'):
+            size = int(alg[3:],0)
+        elif alg.startswith('aes'):
+            size = int(alg[3:],0)
+            # XXX: We set the mode for the tpm to NULL so we can support arbitrary mechanisms on use, but we store it
+            # as CBC for now. We may wish to make this an asterik or NULL indicating we don't filter it out...
+            # Unfortunatley AES mechs are not bitwise flags, so we can't just or them in, and 0 is a valid
+            # CKM_ value, so we can't use 0. Maybe CKM_VENDOR_DEFINED | XXX.
+        else:
+            sys.exit('Unknown algorithm, got: "%s"' % alg)
 
-                # decrypt sobj auth with wrapping
-                sobjauth = sobj['objauth'];
-                if sym_support:
-                    sobjauth = binascii.unhexlify(sobjauth)
-                    sobjauth = tpm2.decrypt(wrappingctx, wrappingauth, sobjauth)
-                else:
-                    c = AESCipher(binascii.unhexlify(wrappingauth))
-                    sobjauth = c.decrypt(sobjauth)
-
-                #create an auth value for the tertiary object.
-                objauth = hash_pass(rand_str(32))['hash']
-
-                # load the secondary object
-                sobjctx = tpm2.load(pobj['handle'], pinpobjauth, sobj['priv'], sobj['pub'])
-
-                # create a tertiary object under the loaded secondary object
-                # Map onto the defaults in tpm2_create and store the attributes into
-                # the db.
-                alg = args['algorithm']
-                if alg.startswith('rsa'):
-                    size = int(alg[3:],0)
-                elif alg.startswith('aes'):
-                    size = int(alg[3:],0)
-                    # XXX: We set the mode for the tpm to NULL so we can support arbitrary mechanisms on use, but we store it
-                    # as CBC for now. We may wish to make this an asterik or NULL indicating we don't filter it out...
-                    # Unfortunatley AES mechs are not bitwise flags, so we can't just or them in, and 0 is a valid
-                    # CKM_ value, so we can't use 0. Maybe CKM_VENDOR_DEFINED | XXX.
-                else:
-                    sys.exit('Unknown algorithm, got: "%s"' % alg)
+        with TemporaryDirectory() as d:
+            tpm2 = Tpm2(d, path)
```

Comment:
Again cant you just use the passed in tpm2?

Reply:
yes